### ISSUE TITLE
present host certificate

### ISSUE THREAD

**Issue by candlerb** [candlerb/sshtokenca#5]
Ensure that sshagentca, if given a host certificate, is able to present it to the client

**Comment by rorycl** [comment]
Do you mean sshagentca should generate host certificates as well as user certificates?

This is obviously pretty easy to do, but ideally would require the provision of a separate private CA key for generating host certificates.

Thinking about it further I think you are suggesting that a previously provided CA public key could be used for checking host certificates presented by clients. It may be simpler to distribute the public key of the server keypair for individuals to add to their `known_hosts` files.

**Comment by candlerb** [comment]
@rorycl:

> Do you mean sshagentca should generate host certificates as well as user certificates?

No: I meant that if sshagentca has been given its own host certificate, it should be able to present it to the client as part of the session establishment.  Then any client who already trusts the host CA can talk to sshagentca without any TOFU issues.

I haven't dug into the details of how to make this work, e.g. whether [ServerConfig.AddHostKey](https://godoc.org/golang.org/x/crypto/ssh#ServerConfig.AddHostKey) can take a certificate.

I was envisaging that host certificates would be issued by a separate host CA, not by sshagentca itself.

-------

There's a separate discussion to be had about what's the best way to issue host certs.  ISTM that issuing host certificates is intimately tied to initial provisioning of the host - as the host needs some way to assert that it is eligible for a certificate with a given principal.

TBH, I don't have a problem with very long-lived host certs; after all, we are usually happy with infinite entries in `~/.ssh/known_hosts`.  The main benefit of host certs is avoiding the initial TOFU exchange for users, and avoiding alerts when a host is rebuilt.

Issuing time-limited host certs every 90 days, LetsEncrypt-style, would be a nice-to-have.

**Comment by rorycl** [comment]
`AddHostKey` appears to allow the addition of private host keys, but keys with the same algorithm will be overwritten, so I'm not sure that is practical.

The practice of distributing host CA keys, like user CA keys, is separate from the role of this server, in my view. All adding a host certificate would do would be to allow users to connect to new/ephemeral hosts without a host verification warning. There seems little point in scoping this by user so it would best be done as a global setting (eg `hosts: *.example.com, www.other.org`). I assume rebuilt/ephemeral servers would have the necessary CA keys added as part of the build process.

Expiry of host certs would best match the user certs in my view. Long-lived client agents should be discouraged, and cycling host and user CA keys on a tight schedule is a good idea. We might have both old and new CA keys for a transition period (which only needs to be as long as the max lifetime of the certs).

**Comment by candlerb** [comment]
> The practice of distributing host CA [public] keys, like user CA keys, is separate from the role of this server,

Absolutely, but I think we are talking at cross purposes.  I am only considering the following situation:

1. You already have a standalone host CA.  (Not sshagentCA).
2. Your users already have the host CA public key, and have enabled it for `*.example.com`
3. You are running sshagentca under hostname `sshagentca.example.com`
4. You want to give sshagentca  a host certificate, signed by the host CA from point (1), which it presents to any user who connects to sshagentca.example.com, to prove its identity.

This is to prevent TOFU warnings when you first connect to sshagentca, and reduces the risk of enabling agent forwarding to a potentially untrusted host.

sshagentca's built-in ssh server currently runs with a host private key - but I couldn't see how you could give it a host private key + host certificate instead.

**Comment by rorycl** [comment]
Thanks for your very cogent notes.

In our use case the sshagentca server would be the issuer of certificates, creating a circular dependency if sshagentca were itself to be verfied by a host key/certificate. I can't think of a good way of resolving that without publishing the an entry to the team to add to their `~/.ssh/known_hosts` files.

I good way around this issue in our team's situation would be to generate, by hand, host certificates  only for sshagentca servers, and register the requisite `@cert-authority` lines in each team member's `~/.ssh/known_hosts` file as a one-off exercise. However, as you previously mentioned, there doesn't seem to be a way of advertising a host CA certificate in go's ssh server along the line of sshd's `HostCertificate`. Maybe testing `AddHostKey` is in order.

My other thought was to deal with this with OTP codes as an ancillary verifier along the lines of my crude sketch for a suggested interface #16 (Suggestion for [an] Auth interface). (I'm imagining aes encrypted otp codes added to the yaml settings file, with the unencryption key provided at server startup). This could help to a degree with host verification, as well as ameliorating the worry of somebody gaining access to a user account with an ssh public key registed with sshagentca.

**Comment by candlerb** [comment]
> Expiry of host certs would best match the user certs in my view.

In other words, you're arguing for highly dynamic issuance of host certs - at least once every 24 hours to be consistent with user certs.

> In our use case the sshagentca server would be the issuer of certificates, creating a circular dependency if sshagentca were itself to be verfied by a host key/certificate.

I hadn't considered how you might use sshagentca to do the issuance of host certs.

Normally these sit in files under `/etc/ssh/`. However I see there's a `HostKeyAgent` setting to sshd.  So you can start an ssh-agent at boot, and sshd can retrieve its host key from that (and presumably a host key + cert pair, as sshagentca issues for user certs)

You would still need some long-term identity which authenticates the host to sshagentca - which could be a static host key, like a traditional host key.  Or some sort of hardware token (with touch presence disabled); or it could use its old cert to receive a new cert.  The latter is moving well into the arena of step-ca, where I don't want to go.

I don't think manually-generated OTP codes will work  for this, if the host needs to *automatically* refresh its host cert every 24 hours.

> to generate, by hand, host certificates only for sshagentca servers, and register the requisite @cert-authority lines in each team member's ~/.ssh/known_hosts file as a one-off exercise.

Certainly, but I'm not sure why you'd issue certs by hand for sshagentca servers but automatically for other servers. What is different about the sshagentca server?

If we're talking abot the user CA instance of sshagentca, it can get its host cert from the host CA the same way any other host would.

If we're talking about the host CA instance of sshagentca, it can sign its own cert.

**Comment by rorycl** [comment]
> I hadn't considered how you might use sshagentca to do the issuance of host certs.

It should be trivially easy to issue host certs by extending `addCertToAgent` to create another certificate with a `Certype: ssh.HostCert`. The hosts list could be scoped by either user or global to the sshagentca server.

> I don't think manually-generated OTP codes will work for this, if the host needs to automatically refresh its host cert every 24 hours.

Our use case would be for users to only connect to sshagentca when about to perform server maintenance.

My further thoughts are as follows:

Assuming an ultimate authority CA keypair for hosts, the ultimate authority CA keys (or certificates) could 1) be added to clients' `known_hosts` as a one-time only action 2) be advertised somehow by sshagentca. 

Normal hosts providing ssh access with host and/or user certificates (provided via sshagentca) would only be configured for sshagentca CA keys, which would be subsidiary to the ultimate authority keypair.

This arrangement, if feasible, wold allow host verification of the sshagentca host/s and fairly rapid CA (host and user) key rotation on normal hosts.

**Comment by candlerb** [comment]
Maybe we're talking at cross purposes again.  It occurs to me there are two different types of host keys (or keys+certificates, but let's just say keys for now):

1. When ssh client A is connecting to sshd server B, there's a host key on B to prove its identity to A to prevent MITM attacks.  The public key is what A caches in its `~/.ssh/known_hosts`.
2. When ssh client A is connecting to sshd server B using [`HostbasedAuthentication`](https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Host-based_Authentication) - the host-to-host trust which is like the old `rsh` / `rlogin`.  In this case, client host A proves its identity to server host B with a host key, and host B trusts host A to say what username is logging in.

Case (1) is what I was thinking of.  If we're using host certificates rather than static host keys in `~/.ssh/known_hosts` then the server needs to have a valid host certificate at all times, since it doesn't know when someone is going to login to it.  Therefore, if the server's certificate is time-limited then it needs to be reissued periodically via a cronjob, and this needs to be fully automated.

Case (2) is really just another form of client authentication, except it's the client *host* that authenticates to the server, rather than the client *user*.  The client host cert is just a special case of user cert.  In this case, the client host cert could be generated on demand, or using an OTP key or whatever, as it's only needed when someone logs in.

Can I confirm we're talking about case (1) and not case (2)?  In other words, we're talking about server host certs, not client host certs?

I've never needed case (2), and host-to-host login trust is pretty antiquated anyway.

**Comment by rorycl** [comment]
> Can I confirm we're talking about case (1) and not case (2)? In other words, we're talking about server host certs, not client host certs?

My apologies for being unclear. I'm certainly talking about the first case (server host certs).

The scenario I was attempting to outline above is that a central authority provides host identification (the term used in `man ssh`, I discover) to sshagentca servers once (with the resulting annoyance for clients) but that all access to other hosts would be identified through sshagentca issued host (and user) certificates. Host "identification" here would be client identification that the host is valid, to protect against MITM attacks.

From a client perspective access to sshagentca hosts would have identification checked by the ultimate authority's host CA key, which will be recorded in each client's `known_hosts`. Identification of normal ssh hosts would be made with sshagentca host certificates signed with a different host CA key to the "ultimate" one, allowing for sshagent CA key rotation.

**Comment by candlerb** [comment]
> From a client perspective access to sshagentca hosts would have identification checked by the ultimate authority's host CA key, which will be recorded in each client's `known_hosts`.

Yes, in the form `@cert-authority *.example.com <host-CA-publickey>`

> Identification of normal ssh hosts would be made with sshagentca host certificates signed with a different host CA key to the "ultimate" one, allowing for sshagent CA key rotation.

Are you saying that SSH CA signatures can be chained?  How is that configured?  I thought the `<host-CA-publickey>` above had to be an actual public key, not "any key signed by this other CA key" - and that to rotate the CA key, each user would have to change their `known_hosts`.

```
@cert-authority *.example.com <old-host-CA-publickey>  # eventually deleted or @revoked
@cert-authority *.example.com <new-host-CA-publickey>
```

**Comment by rorycl** [comment]
I'm attempting to express a chaining mechanism based on:

1. each client user having an `@cert-authority` valid only for sshagentca server/s (to avoid MITM attacks against the sshagentca servers), and 
2. the sshagentca server/s inserting host certificates in client user's forwarded agents to avoid MITM attacks against servers for which the sshagentca server's certificates are valid.

Chaining of actual certificates or pubkeys isn't what I was attempting to describe.

Rory

Edit: the sshagentca CA host from the ultimate authority would need to either be advertised through `AddHostKey` (if that works), or the server private key would need to be somehow signed by the ultimate CA authority. I'm not sure if either is feasible.

Sorry about the edits. Long day.

**Comment by candlerb** [comment]
> I'm attempting to express a chaining mechanism based on:
> 
> 1. each client user having an `@cert-authority` valid only for sshagentca server/s (to avoid MITM attacks against the sshagentca servers), and

OK, it lets you change the host key on the sshagentca server itself without clients giving errors when the key changes [^1]

> 2. the sshagentca server/s inserting host certificates in client user's forwarded agents to avoid MITM attacks against servers for which the sshagentca server's certificates are valid.

Well, the purpose of the user's certificate is to prove the identity of the user.  It's not specifically to avoid MITM attacks (although it is secure against MITM attacks).

There is no chain that I can see here, only two entirely separate CAs:

- one signs server host keys (to prove server identity to client)
- one signs user keys (to prove client identity to server)

These are independent as far as I can see, except in as much as when connecting to sshagentca, the server host key must be validated by the client before the client will accept a user key+cert into its agent.

> Edit: the sshagentca host key would need to either be advertised through `AddHostKey` (if that works), or the server private key would need to be somehow signed by the ultimate CA authority. I'm not sure if either is feasible.

Given that the "AddKey" action to an agent allows insertion of a combined privatekey+certificate, I suspect the same may work for ssh server host keys, but I've not tested it.

-----

[^1] If you have gone to the trouble of setting up a host CA, I don't see why you would restrict yourself to using it only to protect connections to sshagentca.  Connections to other hosts are equally valuable - especially to any other hosts that you are forwarding your agent socket to, or typing passwords into.

**Comment by rorycl** [comment]
Re: 1 above. yes, agreed.

Re 2. I think I've caused a confusion. From my point of view the purpose of sshagentca adding host and user certs is so a client can hit any host with a HostKey/HostCertificate combo signed by sshagentca's CA key. In this way the host certs will help mitigate MITM attacks. The user certs separately provide user auth.

Scribble:

![image](https://user-images.githubusercontent.com/646722/84803362-cf6d8480-aff9-11ea-93c4-ae614c7d6c23.png)

* The UA "ultimate authority" bit at A is completely optional but mitigates MITM attacks between the client and sshagentca B.
* B to C should read : "user and host *certs* added to client agent"
* D & E have their host identities verified by the host certificates provided by B in the client's forwarded agent
* D & E permit (or otherwise) user authentication via user certificates provided by B in the client's forwarded agent.

**Comment by candlerb** [comment]
C logs into D and E using the user key+cert issued by sshagentca to prove the identity of the user.  That is clear.

Servers D and E need the sshagentca public key in order to valid those user certificates.  That is clear.

The parts which are unclear:

B to C: "user and host certs added to client agent".  What host cert are you referring to?  C doesn't need a host cert at all - it's not a server (i.e. it's not running sshd)

"D & E have their host identities verified by the host certificates provided by B in the client's forwarded agent".  This makes no sense to me.  If D & E have host certificates issued by some CA (say "X"), then C needs the _public_ key to X to validate it.  It doesn't get this from the client agent, which only stores private keys.  X's public key would be stored in `~/.ssh/known_hosts` on C, as `@cert-authority *.example.com <pubkey>`

The text between D and E says: "Client connections to D+E identified by host cert in client forwarded agent".  I expect this is saying the same thing, and I still don't understand it.

My understanding of ssh host keys + host certificates is as follows:

* On the server, there is a *private* host key, and optionally a certificate to go with it, usually stored under `/etc/ssh/`
* On the client, `~/.ssh/known_hosts` contains either the matching public key, or a `@cert-authority` assertion to accept any host key signed by the host CA

I don't think the ssh-agent running on the client is involved with the validation of host keys or certificates at all.

**Comment by rorycl** [comment]
I've assumed, but not tested, that a host certificate added to one's forwarded agent would replace the `@cert-authority` line in one's known-hosts file.

Perhaps I've misunderstood the Facebook Engineering\s blog post [Scalable and secure access with SSH](https://engineering.fb.com/security/scalable-and-secure-access-with-ssh/):

> You can also use certificates related to host keys. With that, the SSH clients can automatically trust all hosts with a host certificate signed by the CA, eliminating the need to manually accept every new host you SSH into.

I believe that the configuration for host certificate support on receiving servers (eg D and E) in sshd require `HostCertificate` to point to a public host certificate *and* that the certificate's public key matches a private key specified by `HostKey`. That is a bit more tricky to automate than simply setting `TrustedUserCAKeys` from the CA publickey on B/sshagentca, since one of the server's private keys (HostKeys) would need to be signed by the CA key on B/sshagentca to generate the server's private key and associated public host certificate.

So, I'm hoping if one follows the procedure set out for the [Vault host key signing](https://www.vaultproject.io/docs/secrets/ssh/signed-ssh-certificates#host-key-signing) or [Mike Malone's article about ssh certificates](https://smallstep.com/blog/use-ssh-certificates/), but avoiding the `@cert-authority` step.

I'll have to test and see.

**Comment by candlerb** [comment]
> I've assumed, but not tested, that a host certificate added to one's forwarded agent would replace the `@cert-authority` line in one's known-hosts file.

There seems to be confusion here between "certificate" and "public key of the certificate authority".

A "host certificate" is an assertion of host identity (in a host certificate, the principal is the hostname), with a signature.  The signature is made using the private key of the certificate authority, and verified using the public key of the certificate authority.

As an ssh client, when I connect to an sshd server, as part of the exchange the server sends me its host certificate, and proves that it owns the private key corresponding to the public key in that certificate.

As the client, it's my job to *verify* the host certificate, which I do using the public key of the certificate authority.  Using the single public key of the CA, I can verify *all* host certificates signed by that CA.

Inserting a "host certificate" in my ssh-agent wouldn't help.  Firstly, that would be a single host certificate.  That's not what I want; what I want is the public key of the host CA, which can verify *all* host certificates.  Secondly, ssh-agent only holds private keys (or private keys with the corresponding cert).  Thirdly, my ssh-agent is only accessed when using my private key to identify myself to  the host, not when verifying the host's identity.

> Perhaps I've misunderstood the Facebook Engineering\s blog post [Scalable and secure access with SSH](https://engineering.fb.com/security/scalable-and-secure-access-with-ssh/):
> 
> > You can also use certificates related to host keys. With that, the SSH clients can automatically trust all hosts with a host certificate signed by the CA, eliminating the need to manually accept every new host you SSH into.

That's correct - but the user's ssh-agent does not play a part in this, and this is entirely separate from using user certs to identify users to hosts.  You can deploy user certs without deploying host certs, or vice versa.

If you decide to deploy host certs: the host has a certificate signed by the host CA, and the client has the public key of the host CA, and uses it to verify the identify of the host you're logging into.  If the host has a valid cert, the clients knowns that a trusted third party (the host CA) has already verified the identity of the host, so the user doesn't have to be prompted whether or not to accept the host key.

> I believe that the configuration for host certificate support on receiving servers (eg D and E) in sshd require `HostCertificate` to point to a public host certificate _and_ that the certificate's public key matches a private key specified by `HostKey`.

By "receiving servers" I presume you just mean "servers" (i.e. sshd) that accept incoming connections from clients like C.

You are right: to use host certificates to prove their identity, D and E each have a private/public host key pair and a host certificate.  Let's say the host key is RSA, then the private key, public key and cert are stored in

```
/etc/ssh/ssh_host_rsa_key
/etc/ssh/ssh_host_rsa_key.pub
/etc/ssh/ssh_host_rsa_key-cert.pub   << note
```

respectively.

As part of the exchange, the server proves that it holds the private key corresponding to the public key.

If you're *not* using certificates, then the client verifies that the public key matches the expected public key in `~/.ssh/known_hosts`.

If you *are* using certificates, then the server sends its certificate to the client as well.  The client verifies that the certificate is signed by the host CA, and that the host's public key matches the public key in the certificate, and that the principal in the certificate matches the host's name.

To do this, it needs the host CA's public key as a `@cert-authorit` line in `known_hosts`.

> That is a bit more tricky to automate than simply setting `TrustedUserCAKeys` from the CA publickey on B/sshagentca, since one of the server's private keys (HostKeys) would need to be signed by the CA key on B/sshagentca to generate the server's private key and associated public host certificate.

Signing a certificate does not need to generate a new private key - the CA can sign the existing one - but yes.

Note that the host CA key does not need to be the same as the user CA key.  In general, it's better to keep them completely separate.  `TrustedUserCAKeys` on each server has the user CA public key, and `known_hosts` on each client has the host CA public key.  You could for example use sshagentca to sign user keys, and manual signing of host keys.

You are also correct that automating the signing of host keys is not straightforward.  If you wanted to do it with sshagentca, then you'd need:

1. A cronjob that runs a renewal script periodically
2. The renewal script is an ssh client which connects to sshagentca
3. The renewal script accepts the host cert, intercepts the agentca "AddKey" command which sshagentca sends down the agent session, and instead writes the key+cert to the local filesystem under `/etc/ssh`.  Intercepting the AddKey command is easy, as Go implements the agent protocol as an interface - see https://github.com/candlerb/sshtokenlogin/blob/master/agent.go

Now: if you wanted to be really clever, then *it is possible* on the server to use an instance of ssh-agent to store the host private key and certs. Then, instead of fetching them from `/etc/ssh/ssh_host_rsa_key*` or whatever, sshd would use the agent pointed to by the `HostKeyAgent` setting - it would use the agent to sign the challenge from the client, to prove that the server owns the private key.

This is quite an intrusive change to all your hosts, because you'd need to run a persistent instance of ssh-agent on every host, and modify sshd to talk to it, and prime it with the private key at bootup.  But it *is* symmetrical with the way users prove their identity to the host with ssh-agent.

Personally I wouldn't want to make this change across all my hosts, unless I was using a hardware token for the host key, and I'm not that paranoid.

However: having a special ssh client which fetches host certificates and writes them to `/etc/ssh/` would be an interesting approach for a host CA.

**Comment by rorycl** [comment]
Hi @candlerb 

My confusion has wasted your time -- apologies. I thought adding host certs to client forwarded agents would at least short-circuit the need for the `@cert-authority` entries in `known_hosts`.

The special ssh client for saving host certificates you have suggested sounds pretty interesting. Thanks also for the pointer to AddKey.

**Comment by rorycl** [comment]
Damien Miller kindly responded to my notes about this on the openssh mailing list.

The only feasible element of what I was suggesting might be to "forward around a database of known_hosts, including the @cert-authority markers". However he points out that the agent only handles privat keys, and has no "ancilliary metadata storage and no protocol verbs to perform the
actions needed by host key verification".

Adding these are well outside the scope of the agent and are unlikely to be added to keep the attack surface of the agent as small as possible.
